### PR TITLE
Switch to UV for dependency and venv management

### PR DESCRIPTION
**Pull request by Samuel-Therrien-Beslogic**
~~Depends on #196 because of dependencies updates.~~ Resolving dependencies vulnerabilities as well.

I'm aware of Peotry and pdm, but I'd love supporting Astral in further developing UV by using it ourselves. UV also follows modern Python packaging standards rather than using fully custom solutions (note the concept of a lock file still isn't standardized: https://peps.python.org/pep-0751/ )

This should speedup and simplify dependency and venv management. Making its usage much closer to what we're used to with npm. Whilst also modernizing our dependency declarations.

I'd like a review from someone with more experience with Docker to review

### CODE REVIEW

File: .github/workflows/canopeum_backend.yml [hunk 1]
```diff
@@ -28,13 +30,21 @@ jobs:
...
-      - run: pip install -r requirements-dev.txt
+          path: ${{ env.UV_CACHE_DIR }}
+          key: uv-${{ runner.os }}-${{ hashFiles('uv.lock') }}
+          restore-keys: |
+            uv-${{ runner.os }}-${{ hashFiles('uv.lock') }}
+            uv-${{ runner.os }}
+      - run: uv sync --locked --extra dev
       - run: mypy . --python-version=3.12
+      - name: Minimize uv cache
+        run: uv cache prune --ci
```
Comment (Samuel-Therrien-Beslogic) [inline .github/workflows/canopeum_backend.yml:47]: After consideration, I'll avoid dependency caching for now. The speedup in install from having to redownload is mostly offset by the caching overhead. Avoiding potentially random network failure is nice, but rare. And I'd prefer other maintainers not having to handle maintaining this piece of devops.
https://docs.astral.sh/uv/guides/integration/github/#caching

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -1,35 +1,37 @@
...
-			"dprint.dprint"
-		]
-	}
-}
-	// Configure tool-specific properties.
-	// "customizations": {},
+  // Use 'postCreateCommand' to run commands after the container is created.
+  "postCreateCommand": "pip3 install --user -e canopeum_backend[dev]",
+  // https://github.com/astral-sh/uv/issues/5964
+  // "postCreateCommand": "cd canopeum_backend && uv sync --locked --extra dev",
```
Comment (Samuel-Therrien-Beslogic) [inline .devcontainer/devcontainer.json:17]: Ref https://github.com/astral-sh/uv/issues/5964

File: canopeum_backend/Dockerfile [hunk 3]
```diff
@@ -1,13 +1,15 @@
 FROM python:3.12.2
+COPY --from=ghcr.io/astral-sh/uv:latest /uv /bin/uv
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/Dockerfile:2]: https://docs.astral.sh/uv/guides/integration/docker/#installing-uv

Comment (sam9291) [reply canopeum_backend/Dockerfile:2]: Neat I didn't know we could copy directly from another image, good to know :)

**Pull request merged**